zerver/actions/message_edit.py

```diff
@@ -1152,6 +1152,28 @@ def user_info(um: UserMessage) -> dict[str, Any]:
     )
 
 
+def is_resolve_topic_request(
+    message: Message,
+    new_topic_name: str | None,
+    propagate_mode: str,
+    content: str | None,
+) -> bool:
+    old_topic_name = message.topic_name()
+
+    if new_topic_name is not None:
+        topic_resolved: bool = new_topic_name.startswith(
+            RESOLVED_TOPIC_PREFIX
+        ) and not old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+        topic_unresolved: bool = old_topic_name.startswith(
+            RESOLVED_TOPIC_PREFIX
+        ) and not new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)
+
+        if topic_resolved or topic_unresolved:
+            return True
+
+    return False
```

Comment:
This check doesn't seem right -- or at least, it seems potentially wrong? I think we need to be checking that the topic is identical except for the difference in `RESOLVED_TOPIC_PREFIX`, except slightly more complex thanks to the issue of truncation of very long topic names...

Reply:
Surely we have other logic in this file that would be more suitably reused for this check... yeah why don't you try extracting for reuse the first part of these functions?
```
def maybe_send_resolve_topic_notifications(                                                             
    *,                                                                                                  
    user_profile: UserProfile,                                                                          
    stream: Stream,                                                                                     
    old_topic_name: str,                                                                                
    new_topic_name: str,                                                                                
    changed_messages: QuerySet[Message],                                                                
    pre_truncation_new_topic_name: str,                                                                 
) -> tuple[int | None, bool]:                                                                           
    """Returns resolved_topic_message_id if resolve topic notifications were in fact sent."""           
    # Note that topics will have already been stripped in check_update_message.                         
    resolved_prefix_len = len(RESOLVED_TOPIC_PREFIX)                                                    
    topic_resolved: bool = (                                                                            
        new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)                                                
        and not old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)                                        
        and pre_truncation_new_topic_name[resolved_prefix_len:] == old_topic_name                       
    )                                                                                                   
    topic_unresolved: bool = (                                                                          
        old_topic_name.startswith(RESOLVED_TOPIC_PREFIX)                                                
        and not new_topic_name.startswith(RESOLVED_TOPIC_PREFIX)                                        
        and old_topic_name.lstrip(RESOLVED_TOPIC_PREFIX) == new_topic_name                              
    )                                                                                                   
                                                                                                        

```

Reply:
Maybe some other reorganization is appropriate -- I feel like structurally, we want to just be computing these values once at the start and using those values throughout, not parsing it like 5 times. 

Possibly we should start with a preparatory PR refactoring this code path more deeply, perhaps using a new dataclass (or collection of 3 dataclasses) similar in nature to `MessageSendRequest` that can contain computed state like this and be passed around. Like, a `ResolveTopicRequest` object could have these values, an `MessageContentEditRequest` can have the stuff for editing content only, and a `MoveMessagesRequest` can contain the state ... and we can try to arrange that an initial set of logic does the parsing/checking and generates one of those 3 types, which can then be used with `isinstance` checks and the like to do the rest of the message-edit flow.

Would you be up for attempting some version of that? I feel like structurally, this function is getting unwieldy and breaking it up would be nice, but these tiny helpers with so many parameters don't feel like a nice way to read something like this either.

Reply:
ya, I agree this code path doesn't actually look so good. Let me give that a try!